docs/ir/ir-breaking-changes.rst

```diff
@@ -33,3 +33,38 @@ Consequently, if the padding space within a struct is used to store data (e.g. i
     }
 
 We have the same behavior for implicit delete, for example when array of structs is shortened.
+
+ * The order of contract initialization has changed in case of inheritance.
+
+The order used to be:
+ - Evaluate base constructor arguments from most derived to most base contract.
+ - Initialize all state variables in the whole inheritance hierarchy from most base to most derived.
+ - Run the constructor, if present, for all contracts in the linearized hierarchy from most base to most derived.
+
+New order:
+ - All state variables are zeroes at the beginning
+ - Evaluate base constructor arguments from most derived to most base contract.
+ - For every contract in order from most base to most derived in the linearized hierarchy execute:
+     1. Initialize state variables as 0 or assigned value, if present at declaration.
```

Comment:
Are they really initialized to ``zero`` again here? They are already zero from the first bullet point... so maybe this is rather just
```suggestion
     1. State variables are assigned value their initial values, if present at declaration.
```

I almost said the distinction is not relevant, but actually it is - what does the following do?

```
contract A {
  uint public x;
  constructor(uint) {}
  function f() public { x = 4; }
}
contract B is A {
  constructor() A(f()) {}
}
```

Is ``A.x()`` zero or four here? The current formulation here suggests that first the base arguments are evaluated, settin g ``x`` to ``4``, but then afterwards it's "initialized" back to zero... my suggestion would have it be ``4`` instead...

Also interesting, but independent of the formulation here:


```
contract A {
  uint public x = 5;
  constructor(uint) {}
  function f() public { x = 4; }
}
contract B is A {
  constructor(uint) A(f()) {}
}
contract C is B {
   uint y;
   function g() public { y = x; }
   constructor() B(g()) {}
}
```
Here ``x`` would be ``4`` for the duration of base constructor argument evaluation (after the call to ``f``), so ``y`` would end up being ``4``, but in the end ``x`` will be ``5``, right :-)?

Reply:
So to be clear: if in the first snippet above ``x`` ends up actually being ``zero``, then my point is mute and the PR is fine ;-).

Reply:
I think you are actually right. Adding such test and if ends up as You said will update documentation as well.

Reply:
Yes, the distinction is relevant, and they're not reassigned to 0. When I wrote that the first bullet point wasn't there yet :p